### ISSUE TITLE
Problem with using Sequencer with non-recurrent modules

### ISSUE THREAD

**Issue by kmul00** [Element-Research/rnn#129]
Hi,

I have a model where I have a `CNN` (as an *encoder*) followed by an `LSTM`, acting on a *sequential data*. It can be thought of like, *features* are learnt from my *input* using a `CNN` and these *features* are then fed into an `LSTM` for final *classification*. For ease, I wrapped my modules in a `sequencer` to allow using *sequential* data.

My **input** is as follows :
```lua
input = torch.Tensor(time_step X batch_size X channels X image_width X image_height)
```
My **model** is as follows:
```lua

-- a normal three layer CNN encoder
-- conv -> batch_norm -> activation -> pooling -> dropout

cnn_model = nn.Sequential()
cnn_model:add(SpatialConvolution):add(BatchNormalization):add(ReLU):add(MaxPooling):add(Dropout)
cnn_model:add(SpatialConvolution):add(BatchNormalization):add(ReLU):add(MaxPooling):add(Dropout)
cnn_model:add(SpatialConvolution):add(BatchNormalization):add(ReLU):add(MaxPooling):add(Dropout)
cnn_model:Reshape()  -- to flatten out the features while feeding into LSTM

-- LSTM layer

lstm_model = nn.LSTM(input_size, hidden_size)
lstm_model:add(Dropout)

-- putting everything together

model = nn.Sequential()
model:add(nn.SplitTable(1))  -- sequencer expects table of inputs
model:add(nn.Sequencer(cnn_model))  -- putting the CNN into sequencer
model:add(nn.Sequencer(lstm_model))  -- followed by LSTM which acts on output of the CNN
model:add(nn.SelectTable(-1)) -- many-to-one LSTM
model:add(nn.Linear()):add(nn.LogSoftMax())  -- apply criterion on top of this

criterion = nn.ClassNLLCriterion()
```

Now **training** is okay, error decreasing as expected. Problem arises while **testing**. For verification purpose I am using the same *train set* as my *test set*. 

While **testing** when I set 
```lua
model:evaluate()
```
things are not working out. The accuracy keeps oscillating at a very low value (~5-10%), even though my *training accuracy* is much higher and I am using the same data for *testing* also.

But instead if I use 
```lua
cnn_model:evaluate()
lstm_model:evaluate()
```
things are going perfectly, as expected !

I have previously used `LSTM` without any problem. So my guess is, I am doing something wrong while putting the `CNN` in the `sequencer`.

Can someone point out the reason for this and where exactly is my mistake ? 

Thanks !

**NB**: I have omitted details such as `kernel_size`, `hidden_unit_size`, `linear_unit_size`, *etc*, for ease of readability. There is no problem like dimension mismatch or something of that sort.

**Comment by nicholas-leonard** [comment]
@kmul00 I am trying to reproduce this issue into a unit test in test/test.lua : 
```lua
function rnntest.issue129()
   -- a normal three layer CNN encoder
   -- conv -> batch_norm -> activation -> pooling -> dropout

   local cnn_model = nn.Linear(3,4)

   -- LSTM layer

   local lstm_model = nn.LSTM(4, 4)

   -- putting everything together

   local model = nn.Sequential()
   --model:add(nn.PrintSize"split1")
   model:add(nn.SplitTable(1))  -- sequencer expects table of inputs
   --model:add(nn.PrintSize"split")
   model:add(nn.Sequencer(cnn_model))  -- putting the CNN into sequencer
   model:add(nn.Sequencer(lstm_model))  -- followed by LSTM which acts on output of the CNN
   model:add(nn.SelectTable(-1)) -- many-to-one LSTM
   model:add(nn.Linear(4,4)):add(nn.LogSoftMax())  -- apply criterion on top of this
   
   local input = torch.randn(7, 5, 3) -- seqlen x batchsize x inputsize
   local gradOutput = torch.randn(5, 4)
   
   model:training()
   local output = model:forward(input):clone()
   
   model:evaluate()
   local output2 = model:forward(input):clone()
   
   mytester:assertTensorEq(output, output2,  0.0000001, "issue 129 output err")
end
```

I can't get it to fail. Can you modify it to fail for your use case so I can debug it?

**Comment by kmul00** [comment]
Hi @nicholas-leonard , 

Thanks for looking into it.  Whats is the command to run the tests ?

Is it `th -lnn -e "nn.test{'issue129'}"` ?

**Comment by kmul00** [comment]
@nicholas-leonard 

It seems the problem is with the `nn.Dropout()` module.

The following test fails.

```lua
function rnntest.issue129()
   -- a normal three layer CNN encoder
   -- conv -> batch_norm -> activation -> pooling -> dropout

   local cnn_model = nn.Sequential()
   cnn_model:add(nn.Linear(3, 4))
   cnn_model:add(nn.Dropout(0.5)) -- without this it passes

   -- LSTM layer

   local lstm_model = nn.LSTM(4, 4)
    lstm_model:add(nn.Dropout(0.5)) -- without this it passes

   -- putting everything together

   local model = nn.Sequential()
   --model:add(nn.PrintSize"split1")
   model:add(nn.SplitTable(1))  -- sequencer expects table of inputs
   --model:add(nn.PrintSize"split")
   model:add(nn.Sequencer(cnn_model))  -- putting the CNN into sequencer
   model:add(nn.Sequencer(lstm_model))  -- followed by LSTM which acts on output of the CNN
   model:add(nn.SelectTable(-1)) -- many-to-one LSTM
   model:add(nn.Linear(4,4)):add(nn.LogSoftMax())  -- apply criterion on top of this

   local input = torch.randn(7, 5, 3) -- seqlen x batchsize x inputsize
   local gradOutput = torch.randn(5, 4)

   model:training()
   local output = model:forward(input):clone()
   
   model:evaluate()
   local output2 = model:forward(input):clone()

   mytester:assertTensorEq(output, output2,  0.0000001, "issue 129 output err")
end
```

**Comment by nicholas-leonard** [comment]
@kmul00 It is normal for the test to fail with dropout as it is stochastic during training. Can you make the test fail without it?

**Comment by kmul00** [comment]
Yeah, that's what I realized afterwards, hence deleted that post. Will come up with e test without `dropout`.

**Comment by kmul00** [comment]
@nicholas-leonard , I think I am able to zero in the problem now. It seems the problem lies with **`BatchNormalization`**. Following is the test I used to determine it.

I have converted it to replicate my exact use-case.

```lua
function rnntest.issue129()
   -- a normal three layer CNN encoder
   -- conv -> batch_norm -> activation -> pooling -> dropout

   local cnn_model = nn.Sequential()
   
   -- this one fails (condition 1)
   cnn_model:add(nn.SpatialConvolution(2, 16, 5, 5)):add(nn.SpatialBatchNormalization(16)):add(nn.SpatialMaxPooling(2, 2))
   cnn_model:add(nn.SpatialConvolution(16, 32, 3, 3)):add(nn.SpatialBatchNormalization(32)):add(nn.SpatialMaxPooling(2, 2))
   cnn_model:add(nn.SpatialConvolution(32, 64, 3, 3)):add(nn.SpatialBatchNormalization(64)):add(nn.SpatialMaxPooling(2, 2))
   cnn_model:add(nn.Reshape(64*6*6))
   
   -- this one passes (condition 2)
   --[[cnn_model:add(nn.SpatialConvolution(2, 16, 5, 5)):add(nn.SpatialMaxPooling(2, 2))
   cnn_model:add(nn.SpatialConvolution(16, 32, 3, 3)):add(nn.SpatialMaxPooling(2, 2))
   cnn_model:add(nn.SpatialConvolution(32, 64, 3, 3)):add(nn.SpatialMaxPooling(2, 2))
   cnn_model:add(nn.Reshape(64*6*6))]]--
   
   -- LSTM layer

   local lstm_model = nn.LSTM(64*6*6, 512)
   
   -- putting everything together

   local model = nn.Sequential()
   model:add(nn.SplitTable(1))  -- sequencer expects table of inputs
   model:add(nn.Sequencer(cnn_model))  -- putting the CNN into sequencer
   model:add(nn.Sequencer(lstm_model))  -- followed by LSTM which acts on output of the CNN
   model:add(nn.SelectTable(-1)) -- many-to-one LSTM
   model:add(nn.Linear(512,20)):add(nn.LogSoftMax())  -- apply criterion on top of this

   local input = torch.randn(7, 5, 2, 64, 64) -- seqlen x batchsize x channels x height x width
   
   model:training()
   local output = model:forward(input):clone()

   -- this works fine for both condition 1 and 2
   cnn_model:evaluate()
   lstm_model:evaluate()
   local output2 = model:forward(input):clone()

   mytester:assertTensorEq(output, output2,  0.0000001, "issue 129 cnn, lstm output err")

   -- here it fails for condition 1
   model:evaluate()
   local output3 = model:forward(input):clone()

   mytester:assertTensorEq(output, output3,  0.0000001, "issue 129 model output err")
end
```

**Comment by nicholas-leonard** [comment]
@kmul00 This rnn-less example fails : 

```lua
local sbn = nn.SpatialBatchNormalization(16)
   local input = torch.randn(2,3,8,8)
   
   sbn:training()
   local output = sbn:forward(input):clone()
   sbn:evaluate()
   local output2 = sbn:forward(input):clone()
   mytester:assertTensorEq(output, output2, 0.0000001)
```

So I would venture a guess that the problem lies in how SpatialBatchNormalization is evaluating your test set. @soumith I am guessing this is normal behaviour. Do you have any suggestions?

**Comment by kmul00** [comment]
I have a question. Then why does casting the modules individually doesn't fail ?
After all, the `batchnormalization` modules were inside the `cnn_model`. Why does casting it separately as `cnn_model:evaluate()`, instead of doing the whole `model:evaluate()`, doesn't fail ?

**Comment by nicholas-leonard** [comment]
@kmul00 Because it doesn't actually cast the batch normalization modules. These are cloned internally by the Sequencer (via Recursor). So you are only casting the original. In effect, the clones are still in training() mode.

**Comment by kmul00** [comment]
Oh !
Thanks for the clarification. :)

So what do you suggest ?

Do I get away from using `BatchNormalization` or wrap the model in some other way ?

**Comment by soumith** [comment]
@kmul00 @nicholas-leonard output2 wont be equal to output, because the running_mean and running_var are exponential averages over some momentum.

**Comment by soumith** [comment]
if you run this line enough times:
sbn:training()
   local output = sbn:forward(input):clone()

Then 
   sbn:evaluate()
   local output2 = sbn:forward(input):clone()
   mytester:assertTensorEq(output, output2, 0.0000001)

will pass

**Comment by kmul00** [comment]
@soumith Thanks for the clarification.

@nicholas-leonard The problem is with `BatchNormalization` module, when wrapped in a `sequencer`.

Reason is, while in `evaluate()` mode, the **running_mean** and **running_std** learnt during `training()` are getting lost and it's falling back to it's default value of `running_mean = 0` and `running_std = 1`. I have actually printed the values and verified it.

The following tests further verifies it. 
Below is the test for normal `BatchNormalization` module, modified according to @soumith 's suggestion.
```lua
function rnntest.issue129()
   
   local model = nn.Sequential()
   model(nn.SpatialBatchNormalization(2))
   
   local input = torch.randn(4, 2, 64, 64)

   model:training()
   local output
   for i=1, 200 do  -- to run training enough times
      output = model:forward(input):clone()
   end

   model:evaluate()
   local output2 = model:forward(input):clone()

   mytester:assertTensorEq(output, output2,  0.0000001, "issue 129 err")
end
```
This one passes without any error.

But if we convert the above code to wrap the model in a `sequencer` like below,
```lua
function rnntest.issue129()
   
   local normalize = nn.Sequential()
   normalize:add(nn.SpatialBatchNormalization(2))
   
   model = nn.Sequential()
   model:add(nn.SplitTable(1))
   model:add(nn.Sequencer(normalize))
   model:add(nn.JoinTable(1))

   local input = torch.randn(1, 4, 2, 64, 64)

   model:training()

   local output
   for i=1, 200 do
      output = model:forward(input):clone()
   end

   model:evaluate()
   local output2 = model:forward(input):clone()

   mytester:assertTensorEq(output, output2,  0.0000001, "issue 129 err")
end
```
the test fails.

Is there a way to modify the code to keep track of the running_mean and running_std learnt during training, while evaluating ?

Any head-starts regarding which files to look into (`sequencer.lua`, `recursor.lua`, e.t.c) will also be very helpful. I can give it a try then.

Thanks !

**Comment by nicholas-leonard** [comment]
@kmul00 I included your very helpful unit test and fixed it in https://github.com/Element-Research/rnn/commit/c03029ac3d4299e718a4adac554fb9b545a091bc . The precision isn't as good as in your original unit test. But your original non-rnn unit test was also failing when I fixed `model(nn.Spatial...)`. So I reduced the precision of the test to 0.0002. I hope this works for you. Basically the issue was the Recursor wasn't adding its recurrentModule to the sharedClones.
